job-server/src/spark.jobserver/auth/LdapGroupRealm.scala

```diff
@@ -0,0 +1,142 @@
+ * the 'member' attribute values of each group against the given user.
+ *
+ * @note not all LDAP installations use the member property.... we might have to add
+ *   memberOf matches as well and others
+ *
+ * @author KNIME (basics in Java stem from different sources by various authors from stackoverflow and such)
+ */
+class LdapGroupRealm extends JndiLdapRealm {
+  import collection.JavaConverters._
+
+  private val logger = LoggerFactory.getLogger(getClass)
+
+  private val searchCtls: SearchControls = {
+    val c = new SearchControls();
+    c.setSearchScope(SearchControls.SUBTREE_SCOPE);
+    c
+  }
+
+  lazy val searchBase: String = getContextFactory() match {
+    case jni: JndiLdapContextFactory =>
+      getEnvironmentParam(jni, "ldap.searchBase")
+    case _ =>
+      throw new RuntimeException("Configuration error: " +
+        "LdapGroupRealm requires setting of the parameter 'searchBase'")
+  }
+
+  lazy val allowedGroups: Option[Array[String]] = getContextFactory() match {
+    case jni: JndiLdapContextFactory =>
+      val groups = getEnvironmentParam(jni, "ldap.allowedGroups").
+        split("\"").filter(group => group.replaceAll("[, ]", "").length() > 0)
+      if (groups.isEmpty) {
+        None
+      } else {
+        Some(groups)
+      }
+    case _ =>
+      None
+  }
+
+  private def getEnvironmentParam(jni: JndiLdapContextFactory, param: String): String = {
+    val value = jni.getEnvironment().get(param)
+    value match {
+      case null =>
+        //tell user what is missing instead of just throwing a NPE
+        throw new RuntimeException("Configuration error: " +
+          "LdapGroupRealm requires setting of the parameter '" + param + "'")
+      case v =>
+        v.toString
+    }
+  }
+
+  override def queryForAuthorizationInfo(principals: PrincipalCollection,
+                                         ldapContextFactory: LdapContextFactory): AuthorizationInfo = {
+
+    val username = getAvailablePrincipal(principals).toString
+    val ldapContext = ldapContextFactory.getSystemLdapContext()
+    try {
+      queryForAuthorizationInfo(ldapContext, username)
+    } finally {
+      LdapUtils.closeContext(ldapContext);
+    }
+  }
+
+  def queryForAuthorizationInfo(ldapContext: LdapContext, username: String): AuthorizationInfo = {
+    val roleNames = getRoleNamesForUser(ldapContext, username)
+    if (isInAllowedGroupOrNoCheckOnGroups(roleNames)) {
+      new SimpleAuthorizationInfo(roleNames.asJava)
+    } else {
+      throw new AuthorizationException(LdapGroupRealm.ERROR_MSG_NO_VALID_GROUP)
+    }
+  }
+
+  //TODO - can (should?) this information be cached?
+  def retrieveGroups(ldapContext: LdapContext): Map[String, Set[String]] = {
+    logger.trace("Retrieving group memberships.")
+
+    val groupSearchAtts: Array[Object] = Array()
+
+    val groupAnswer = ldapContext.search(searchBase, LdapGroupRealm.groupMemberFilter,
+      groupSearchAtts, searchCtls).asScala
+
+    groupAnswer.map(sr2 => {

```

Comment:
Minor nit:  Stylistically the idiom for Scala is usally like this:

    groupAnswer.map { sr2 =>
      logger.debug(....)
    }.toMap